In [1]:
import pandas as pd, numpy as np
df = pd.read_csv('https://raw.githubusercontent.com/IndraniMandal/ds-assets/main/assets/messy_covid19_southamerica%20-%20covid19_southamerica.csv')

In [ ]:
df.shape

(23150, 7)

In [ ]:
df.head()


ObservationDate Country_Region Province_State  Confirmed Deaths Recovered  \
0      2020-01-23         Mexico            NaN        0.0      0         0   
1      2020-01-23         Brazil            NaN        0.0      0         0   
2      2020-01-23       Colombia            NaN        0.0      0         0   
3      2020-02-26         Brazil            NaN        1.0      0         0   
4      2020-02-26         Brazil            NaN        1.0      0         0   

   Active  
0     0.0  
1     0.0  
2     0.0  
3     1.0  
4     1.0

In [ ]:
df.isnull().sum()

ObservationDate       0
Country_Region        0
Province_State     5989
Confirmed             1
Deaths                3
Recovered             2
Active               26
dtype: int64

Question:Which column has the maximum missing value?

Answer: The "Province State" is the column with the maximum missing value.

In [ ]:
percentage_missing = (df['Province_State'].isnull().sum() / len(df)) * 100
print(f"Percentage of missing values in 'Province_State': {percentage_missing:.2f}%")


Percentage of missing values in 'Province_State': 25.87%


Question: What % value is missing in that column?

Answer: Percentage of missing values in 'Province_State': 25.87%

Question:Should you keep this column or drop it? Why?

Answer: I should keep this column because the percentage of value sthat is missing isn't high and the column has 132 unique information which can be valuable information

In [ ]:
print(f"Unique values in 'Province_State': {df['Province_State'].nunique()}")
print(f"Unique values in 'Confirmed': {df['Confirmed'].nunique()}")
print(f"Unique values in 'Deaths': {df['Deaths'].nunique()}")
print(f"Unique values in 'Recoverd': {df['Recovered'].nunique()}")
print(f"Unique values in 'Active': {df['Active'].nunique()}")

Unique values in 'Province_State': 133
Unique values in 'Confirmed': 13583
Unique values in 'Deaths': 4258
Unique values in 'Recoverd': 9801
Unique values in 'Active': 9549


Question: There are some columns with only a few data points missing. Should you drop these columns too? Why or why not?

Answer: I should keep the other columns that have a few missing data points because they have a lot of unique information and also they are significant columns that are important for identifying.

Question: How should you handle these specific data points(particular row)?

Answer: I will handle the specific data points that are missing by leaving them as they are.

Question: There are some data that has incorrect information like instead of a number they might have strings. How are you handling these data points(rows)?

Answer: If it has incorrect information in that row I will impute it with a mean value

In [10]:
df['Province_State'] = pd.to_numeric(df['Province_State'], errors='coerce')

#fill NaN values with the mean
df['Province_State'].fillna(df['Province_State'].mean(), inplace=True)
df.Province_State[df.Province_State.isnull()]

<ipython-input-10-7aea4a1d5922>:4: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['Province_State'].fillna(df['Province_State'].mean(), inplace=True)


Series([], Name: Province_State, dtype: float64)

In [5]:
df['Confirmed'].fillna(df['Confirmed'].mean(), inplace=True)
df.Confirmed[df.Confirmed.isnull()]


<ipython-input-5-0bd287cd3165>:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['Confirmed'].fillna(df['Confirmed'].mean(), inplace=True)


Series([], Name: Confirmed, dtype: float64)

In [7]:
df['Deaths'] = pd.to_numeric(df['Deaths'], errors='coerce')

#fill NaN values with the mean
df['Deaths'].fillna(df['Deaths'].mean(), inplace=True)
df.Deaths[df.Deaths.isnull()]

<ipython-input-7-af6c88c84aa8>:5: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['Deaths'].fillna(df['Deaths'].mean(), inplace=True)


Series([], Name: Deaths, dtype: float64)

In [8]:
df['Recovered'].fillna(df['Recovered'], inplace=True)
df.Recovered[df.Recovered.isnull()]

10    NaN
41    NaN
Name: Recovered, dtype: object

In [9]:
df['Active'].fillna(df['Active'], inplace=True)
df.Active[df.Active.isnull()]

41      NaN
11507   NaN
12819   NaN
15279   NaN
18463   NaN
19441   NaN
19604   NaN
19767   NaN
19930   NaN
20093   NaN
20256   NaN
20419   NaN
20582   NaN
20745   NaN
20908   NaN
21071   NaN
21397   NaN
21560   NaN
21723   NaN
21886   NaN
22050   NaN
22214   NaN
22378   NaN
22542   NaN
22706   NaN
22870   NaN
Name: Active, dtype: float64

In [ ]:
top_10_countries = df.groupby('Country_Region')['Confirmed'].sum().nlargest(10)
print(top_10_countries)

Country_Region
Brazil                316795070.0
Peru                   56307827.0
Mexico                 50251103.0
Colombia               41215763.0
Chile                  40878456.0
Argentina              29496420.0
Ecuador                11307645.0
Bolivia                 8860657.0
Dominican Republic      8115060.0
Panama                  7804400.0
Name: Confirmed, dtype: float64


In [ ]:
cleaned_df = df.copy()
print(cleaned_df.head())
cleaned_df.to_csv('cleaned_covid19_southamerica.csv', index=False)


  ObservationDate Country_Region Province_State  Confirmed Deaths Recovered  \
0      2020-01-23         Mexico              0        0.0      0         0   
1      2020-01-23         Brazil              0        0.0      0         0   
2      2020-01-23       Colombia              0        0.0      0         0   
3      2020-02-26         Brazil              0        1.0      0         0   
4      2020-02-26         Brazil              0        1.0      0         0   

   Active  
0     0.0  
1     0.0  
2     0.0  
3     1.0  
4     1.0  


In [ ]:
from google.colab import files
files.download('cleaned_covid19_southamerica.csv')


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>